src/Dapr.PluggableComponents/Components/StateStore/StateStoreTransactOperation.cs

```diff
@@ -0,0 +1,86 @@
+/// operation properties are not directly exposed, but instead accessed by one of the Visit() methods.
+/// </remarks>
+public abstract record StateStoreTransactOperation
+{
+    /// <summary>
+    /// "Visits" the operation, calling the appropriate visitor method based on the operation type.
+    /// </summary>
+    /// <param name="onDeleteRequest">Called when the operation is a delete request.</param>
+    /// <param name="onSetRequest">Called when the operation is a set request.</param>
+    public void Visit(Action<StateStoreDeleteRequest> onDeleteRequest, Action<StateStoreSetRequest> onSetRequest)
+    {
+        this.Visit(
+            request =>
+            {
+                onDeleteRequest(request);
+
+                return true;
+            },
+            request =>
+            {
+                onSetRequest(request);
+
+                return true;
+            });
+    }
+
+    /// <summary>
+    /// "Visits" the operation, calling the appropriate visitor method based on the operation type.
+    /// </summary>
+    /// <typeparam name="TReturn">The type of value being returned from the visit.</typeparam>
+    /// <param name="onDeleteRequest">Called when the operation is a delete request.</param>
+    /// <param name="onSetRequest">Called when the operation is a set request.</param>
+    /// <returns>The value returned from the called visitor method.</returns>
+    public abstract TReturn Visit<TReturn>(Func<StateStoreDeleteRequest, TReturn> onDeleteRequest, Func<StateStoreSetRequest, TReturn> onSetRequest);
+
+    internal static StateStoreTransactOperation FromTransactionalStateOperation(TransactionalStateOperation operation)
+    {
+        return operation.RequestCase switch
+        {
+            TransactionalStateOperation.RequestOneofCase.Delete => new StateStoreTransactDeleteOperation(StateStoreDeleteRequest.FromDeleteRequest(operation.Delete)),
+            TransactionalStateOperation.RequestOneofCase.Set => new StateStoreTransactSetOperation(StateStoreSetRequest.FromSetRequest(operation.Set)),
+            _ => throw new ArgumentOutOfRangeException(nameof(operation.RequestCase), String.Format(CultureInfo.CurrentCulture, "The operation type '{0}' is not recognized.", operation.RequestCase))
+        };
+    }
+}
+
+internal sealed record StateStoreTransactDeleteOperation(StateStoreDeleteRequest Request)
+    : StateStoreTransactOperation
+{
+    public override TReturn Visit<TReturn>(Func<StateStoreDeleteRequest, TReturn> onDeleteRequest, Func<StateStoreSetRequest, TReturn> onSetRequest)
+    {
+        return onDeleteRequest(this.Request);
+    }
+}
+
+internal sealed record StateStoreTransactSetOperation(StateStoreSetRequest Request)
+    : StateStoreTransactOperation
+{
+    public override TReturn Visit<TReturn>(Func<StateStoreDeleteRequest, TReturn> onDeleteRequest, Func<StateStoreSetRequest, TReturn> onSetRequest)
+    {
+        return onSetRequest(this.Request);
+    }
+}
```

Comment:
Curious: why you need these 2 types?

Reply:
Same question. I didn't understand how the user will use the`Visit` method, would you mind to do a brief explanation?

Reply:
The transact request needs to expose a sequence of operations, that can (currently) be one of two types (i.e. delete and set).  Those operations will typically be iterated over in sequence for processing.  @tmacam and I went back and forth about this in the [equivalent PR for the Java SDK](https://github.com/dapr-sandbox/components-java-sdk/pull/17).  The ultimate goal and concern was that developers understood which operation types they needed to handle, and to do so in a way that guarded against changes in the future.  (E.g. requiring explicit casts tend to make for "soft contracts" that are easily broken.) 

In the Java case, a "marker operation interface" was added directly to the delete and set request types, so that they could be exposed as a sequence of operations.  The transact request then exposes a "visitor" method that iterates over the sequence and calls the appropriate visitor callback in turn.

I went a slightly different way in the case of .NET, both because I wasn't enthusiastic about the marker interface as well as "ForEach()"-style methods being somewhat discouraged.  Instead, the specific operation requests are exposed in small, opaque wrapper types based on a common, abstract operation type (`StateStoreTransactOperation`).  That abstract type has a `Visit()` method that forces the developer to handle each of the possible operations.

A typical use might look like:

```csharp
        foreach (var operation in request.Operations)
        {
            await operation.Visit(
                async deleteRequest =>
                {
                    this.logger.LogInformation("Processing delete request...");

                    await Task.Delay(100);
                },
                async setRequest =>
                {
                    this.logger.LogInformation("Processing set request...");

                    await Task.Delay(100);
                });
        }
```

This allows use of the more "native" `foreach` sequence iteration style.

Reply:
Understood, thanks